org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jdi/JdiInvocationSynchronizer.java

```diff
@@ -0,0 +1,49 @@
+package scala.tools.eclipse.contribution.weaving.jdt.jdi;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+
+import scala.tools.eclipse.contribution.weaving.jdt.ScalaJDTWeavingPlugin;
+
+/**
+ * Used to execute sync operation(s) on JDI object(s).
+ */
+public final class JdiInvocationSynchronizer {
+  private final static JdiInvocationSynchronizer INSTANCE = new JdiInvocationSynchronizer();
+  public final static long DEFAULT_TIMEOUT_IN_SEC = 10;
+
+  private JdiInvocationSynchronizer() {
+    // no need to instantiate outside
+  }
+
+  private final ReentrantLock jdiLock = new ReentrantLock();
```

Comment:
On second thought: why not use `synchronized`? Synchronized blocks are guaranteed to be released, even if exceptions are thrown, so maybe that would simplify (also thinking about the actor code, and Futures that were needed for releasing the lock)